src/BucketBrigade.qs

```diff
@@ -0,0 +1,128 @@
+    open Microsoft.Quantum.Diagnostics;
+
+///////////////////////////////////////////////////////////////////////////
+// PUBLIC API
+///////////////////////////////////////////////////////////////////////////
+
+    /// # Summary
+    /// Creates a QRAM type corresponding to a bit encoded Bucket Brigade scheme.
+    /// # Input
+    /// ## dataValues
+    /// The data to be stored in the memory.
+    /// ## memoryRegister
+    /// The register that you want to be initialized with the provided data.
+    /// # Output
+    /// An instance of the QRAM type that will allow you to use the memory.
+    operation BucketBrigadeQRAMOracle(dataValues : MemoryCell[], memoryRegister : MemoryRegister) : QRAM {
+        let bank = GeneratedMemoryBank(dataValues);
+
+        for (cell in bank::DataSet) {
+            BucketBrigadeWrite(memoryRegister, cell);
+        }
+
+        return Default<QRAM>()
+            w/ Read <-  BucketBrigadeRead(_, _, _)
+            w/ Write <- BucketBrigadeWrite(_, _)
+            w/ AddressSize <- bank::AddressSize
+            w/ DataSize <- bank::DataSize;
+    }
+
+///////////////////////////////////////////////////////////////////////////
+// INTERNAL IMPLEMENTATION
+///////////////////////////////////////////////////////////////////////////
+    
+    /// # Summary
+    /// Writes a single bit of data to the memory.
+    /// # Input
+    /// ## memoryRegister
+    /// Register that represents the memory you are writing to.
+    /// ## dataValue
+    /// The tuple of (address, data) that you want written to the memory.
+    operation BucketBrigadeWrite(
+        memoryRegister : MemoryRegister, 
+        dataCell :  MemoryCell
+    ) 
+    : Unit {
+        let (address, data) = (dataCell::Address, dataCell::Value);
+        let range = SequenceI (address * Length(data), (address + 1) * Length(data) - 1);
+        ResetAll(Subarray(range, memoryRegister!));
+        ApplyPauliFromBitString(PauliX, true, data, Subarray(range, memoryRegister!));
+    }
+
+    /// # Summary
+    /// Reads out a value from a MemoryRegister to a target qubit given an address.
+    /// # Input
+    /// ## addressRegister
+    /// The qubit register that represents the address to be queried.
+    /// ## memoryRegister
+    /// The qubit register that represents the memory you are reading from.
+    /// ## target
+    /// The qubit that will have the memory value transferred to.
+    operation BucketBrigadeRead(
+        addressRegister : AddressRegister, 
+        memoryRegister : MemoryRegister, 
+        targetRegister : Qubit[]
+    ) 
+    : Unit is Adj + Ctl {
+        using (auxRegister = Qubit[2^Length(addressRegister!)]) {
+            within {
+                X(Head(auxRegister));
+                ApplyAddressFanout(addressRegister, auxRegister);
+            }
+            apply {
+                ReadoutMemory(memoryRegister, auxRegister, targetRegister);
+            }
+        } 
+    }
+
+    operation ReadoutMemory(
```

Comment:
This operation is missing docs; also should it maybe be changed to `internal`? Is it used outside of BucketBrigadeRead?